### PR TITLE
Optimize MapFilterFunction by generating bytecode

### PR DESCRIPTION
**Pull request by geraint0923**

### CODE REVIEW

**Comment by geraint0923** [comment]
Addressed the comments.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java [hunk 1]
```diff
@@ -81,29 +108,94 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
...
+        BytecodeNode loadKeyElement;
+        if (!keyType.equals(UNKNOWN)) {
+            loadKeyElement = new IfStatement()
+                    .condition(block.invoke("isNull", boolean.class, position))
+                    .ifTrue(keyElement.set(constantNull(keyJavaType)))
+                    .ifFalse(keyElement.set(keySqlType.getValue(block, position).cast(keyJavaType)));
         }
-        return resultBuilder.build();
+        else {
+            loadKeyElement = new BytecodeBlock().append(keyElement.set(constantNull(keyJavaType)));
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java]: map key can be unknown type. But map keys cannot be null.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java [hunk 2]
```diff
@@ -81,29 +108,94 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
...
+
+        body.append(new ForLoop()
+                .initialize(position.set(constantInt(0)))
+                .condition(lessThan(position, positionCount))
+                .update(incrementVariable(position, (byte) 2))
+                .body(new BytecodeBlock()
+                        .append(loadKeyElement)
+                        .append(loadValueElement)
+                        .append(keep.set(function.invoke("invokeExact", Boolean.class, keyElement, valueElement)))
+                        .append(new IfStatement("if (keep != null) ...")
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java]: I believe it's easier to write 
```
if (Boolean.TRUE.equals(keep)) {
  ...
}
```
than
```
if (keep != null) {
  if (keep.booleanValue()) {
    ...
  }
}
```
Use `BytecodeExpressions.getStatic`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java [hunk 3]
```diff
@@ -81,29 +108,94 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
...
+                                .condition(notEqual(keep, constantNull(Boolean.class)))
+                                .ifTrue(new IfStatement("if (keep) ...")
+                                        .condition(keep.cast(boolean.class))
+                                        .ifTrue(new BytecodeBlock()
+                                                .append(keySqlType.invoke("appendTo", void.class, block, position, blockBuilder))
+                                                .append(valueSqlType.invoke("appendTo", void.class, block, add(position, constantInt(1)), blockBuilder)))))));
+
+        // blockBuilder.build()
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java]: remove line

**Comment by geraint0923** [comment]
Addressed the comments.

**Comment by wenleix** [comment]
This addresses https://github.com/prestodb/presto/issues/7334

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java (hunk 1 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MapFilterFunction.java]: I think you missed this comment.

**Pull request merged**